### PR TITLE
refactor: Remove direction from constrained step

### PR DESCRIPTION
**Pull request by andiwand**
With this PR I want to remove the propagation direction sign inside the `ConstrainedStep` and put it into the stepper as it feels like additional burden and logic that should not be handled by the `ConstrainedStep`. This will not change anything about the overstepping mechanics.

### CODE REVIEW

**Comment by paulgessinger** [comment]
@HadrienG2

**Comment by HadrienG2** [comment]
From memory, part of the motivation for using signed step sizes was that negative steps naturally comes up when doing backwards navigation. If that is not the case already, you may want to ensure we do have test coverage for this use case.

**Comment by andiwand** [comment]
> From memory, part of the motivation for using signed step sizes was that negative steps naturally comes up when doing backwards navigation. If that is not the case already, you may want to ensure we do have test coverage for this use case.

that also makes sense. my thought was why we need 2 signs, one for overstepping and one for direction, when one is enough. and we put the direction in different places where the mechanism actually does not depend on it. like here the step size can be expressed in the forward direction and then you just flip sign for the backward direction

File: Core/include/Acts/Propagator/ConstrainedStep.hpp [hunk 1]
```diff
@@ -56,12 +54,8 @@ class ConstrainedStep {
   constexpr ConstrainedStep() = default;
 
   /// constructor from Scalar
-  /// navigation direction is inferred by the sign of the step size
   /// @param value is the user given initial value
-  constexpr explicit ConstrainedStep(Scalar value) {
-    m_values[user] = std::abs(value);
-    m_direction = Direction::fromScalar(value);
-  }
+  constexpr explicit ConstrainedStep(Scalar value) { m_values[user] = value; }
```
Comment (HadrienG2) [inline Core/include/Acts/Propagator/ConstrainedStep.hpp:58]: Since step size values are not supposed to always be positive, it might make sense to have assertions around checking that whenever something sets a step size value.

**Comment by andiwand** [comment]
yes I totally agree - my changes break almost all the propagation tests right now so I will try to fix them and get to a state where it does exactly what it did before. I _think_ the existing test cases cover all we need to be sure it works? but I am happy to add additional if you have suggestions

File: Core/include/Acts/Propagator/ConstrainedStep.hpp (hunk 1 above)
Comment (andiwand) [reply Core/include/Acts/Propagator/ConstrainedStep.hpp:58]: my though was that you might intentionally create an object that is in back stepping mode. not sure if that is used or useful tho

Comment (HadrienG2) [reply Core/include/Acts/Propagator/ConstrainedStep.hpp:58]: Wait, I thought back stepping mode was now supposed to be handled via state.stepping.navDir?

Comment (andiwand) [reply Core/include/Acts/Propagator/ConstrainedStep.hpp:58]: no sorry for the confusion - I would keep the back stepping sign but remove. I would only remove the propagation direction sign because the `ConstrainedStep` is invariant for forward or backward

**Comment by andiwand** [comment]
after fixing the overstepping the unit tests are green. physmon is also passing which gives me confidence that it works now. hashes are changing which I did not expect

I will check if I can find the changing numbers in the root files.

but generally ready for review IMO

File: Core/include/Acts/Propagator/AtlasStepper.hpp [hunk 2]
```diff
@@ -1224,7 +1224,7 @@ class AtlasStepper {
            std::abs((C1 + C6) - (C3 + C4)));
       if (std::abs(EST) > std::abs(state.options.tolerance)) {
         h = h * .5;
-        state.stepping.stepSize.setValue(h);
+        state.stepping.stepSize.setValue(h * state.stepping.navDir);
```
Comment (niermann999) [inline Core/include/Acts/Propagator/AtlasStepper.hpp:1227]: I guess this is to remove any sign again?

File: Tests/UnitTests/Core/Propagator/AtlasStepperTests.cpp [hunk 3]
```diff
@@ -551,12 +551,12 @@ BOOST_AUTO_TEST_CASE(StepSize) {
   // TODO figure out why this fails and what it should be
   // BOOST_CHECK_EQUAL(stepper.overstepLimit(state), tolerance);
 
-  stepper.setStepSize(state, 5_cm);
-  BOOST_CHECK_EQUAL(state.previousStepSize, navDir * stepSize);
-  BOOST_CHECK_EQUAL(state.stepSize.value(), 5_cm);
+  stepper.setStepSize(state, 5_mm);
```
Comment (niermann999) [inline Tests/UnitTests/Core/Propagator/AtlasStepperTests.cpp:554]: Is the unit switch just cleanup?

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp [hunk 4]
```diff
@@ -264,12 +264,12 @@ BOOST_AUTO_TEST_CASE(eigen_stepper_test) {
   // Step size modifies
   const std::string originalStepSize = esState.stepSize.toString();
 
-  es.setStepSize(esState, 1337.);
-  BOOST_CHECK_EQUAL(esState.previousStepSize, navDir * stepSize);
-  BOOST_CHECK_EQUAL(esState.stepSize.value(), 1337.);
+  es.setStepSize(esState, 17.);
```
Comment (niermann999) [inline Tests/UnitTests/Core/Propagator/StepperTests.cpp:267]: Same here: why is this needed?

File: Core/include/Acts/Propagator/AtlasStepper.hpp (hunk 2 above)
Comment (andiwand) [reply Core/include/Acts/Propagator/AtlasStepper.hpp:1227]: yeah exactly this is supposed to drop the sign we applied before again

File: Tests/UnitTests/Core/Propagator/AtlasStepperTests.cpp (hunk 3 above)
Comment (andiwand) [reply Tests/UnitTests/Core/Propagator/AtlasStepperTests.cpp:554]: don't remember but I guess this is arbitrary. will try to revert

File: Tests/UnitTests/Core/Propagator/StepperTests.cpp (hunk 4 above)
Comment (andiwand) [reply Tests/UnitTests/Core/Propagator/StepperTests.cpp:267]: don't remember but I guess this is arbitrary. will try to revert

**Comment by andiwand** [comment]
[kf_diff.txt](https://github.com/acts-project/acts/files/12096306/diff.txt)

checked the diff looks like noise. will update the hashes

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180034) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180033) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180032) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180031) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsTrkSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180030) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180029) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180028) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180027) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180026) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180025) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180024) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180023) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180022) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180021) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180020) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180019) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31180018) |
